File: 04_callingcard_analysis/CC_peak_overlap_venn.R
## =====================================================================
## Male vs Female Calling-Card peak overlap  ->  Venn + corrected peak sets
##
## WHY THIS SCRIPT EXISTS
##   The previously committed sets (male_unique_peaks_originalPeaks.bed 7,357 /
##   female_unique_peaks_originalPeaks.bed 7,495 / shared_peaks_originalPeaks.bed
##   3,187) were produced by a notebook that is not in this repo, and they do not
##   reproduce from the raw per-sex calls. Audit (2026-07-17) found:
##     - 1,433 male + 1,090 female raw peaks silently dropped
##       (7,357+3,187 = 10,544 vs 11,977 raw male; 7,495+3,187 = 10,682 vs 11,772 raw female)
##     - 997 "male_unique" peaks in fact overlap a raw female peak (and vice versa)
##     - only ~2,122/3,187 "shared" rows exactly match a raw peak despite the
##       "originalPeaks" filename -> coordinates had been merged
##     - a single symmetric shared count (3,187) is not obtainable from a
##       peak-centric intersect, which is inherently many-to-many/asymmetric
##   This script regenerates the split reproducibly from the raw calls.
##
## bedtools NOTE
##   bedtools is not installed on this Windows box. GenomicRanges::findOverlaps
##   with default settings is semantically identical to `bedtools intersect`
##   (>= 1 bp overlap on the same chromosome, strand-agnostic). The equivalent
##   bedtools calls are given per block below so this can be checked on HTCF.
##
## TWO WAYS TO COUNT (both reported; they answer different questions)
##   (1) PEAK-CENTRIC  = `bedtools intersect -u/-v`, keeps each sex's own peaks.
##       Honest but the two "shared" counts DISAGREE (3,712 male-side vs 3,719
##       female-side) because overlap is many-to-many. Cannot draw a 2-circle Venn.
##   (2) REGION-CENTRIC = merge the union, then ask which sexes hit each region.
##       Categories are disjoint and sum exactly -> this is what the Venn uses.
##
## Inputs : raw per-sex MACCs calls (pycallingcards), unaffected by the bug above.
## Outputs: 04_callingcard_analysis/output/peak_overlap/
## =====================================================================

suppressMessages({
  library(GenomicRanges)
  library(VennDiagram)
})
futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")  # silence per-plot logs

## ---- Paths ----------------------------------------------------------
cc_dir  <- "C:/Users/loril/Documents/Egr1/Egr1CC_vs_Egr1KDBulkRNA_FINAL_July2025"
out_dir <- "C:/Users/loril/Documents/GitHub/Egr1-Llaci/04_callingcard_analysis/output/peak_overlap"
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)

male_file   <- file.path(cc_dir, "Male_Egr1CC_peaks_072825.txt")
female_file <- file.path(cc_dir, "Female_Egr1CC_peaks_072825.txt")
stopifnot(file.exists(male_file), file.exists(female_file))

read_cc <- function(f) {
  d <- read.delim(f, sep = "\t", check.names = FALSE, stringsAsFactors = FALSE)
  gr <- GRanges(d$Chr, IRanges(d$Start, d$End))
  mcols(gr)$gene <- d$`Gene Name1`
  mcols(gr)$dist <- d$Distance1
  sort(gr)
}
write_bed <- function(gr, f) {
  write.table(data.frame(chr = as.character(seqnames(gr)), start = start(gr), end = end(gr)),
              file.path(out_dir, f), sep = "\t", quote = FALSE,
              row.names = FALSE, col.names = FALSE)
}

M <- read_cc(male_file)
F <- read_cc(female_file)
cat(sprintf("Raw peaks:  male %d   female %d\n\n", length(M), length(F)))

## ---- (1) PEAK-CENTRIC intersect -------------------------------------
## bedtools intersect -a male.bed -b female.bed -v   (unique)
## bedtools intersect -a male.bed -b female.bed -u   (shared, male coords)
m_hit <- overlapsAny(M, F)
f_hit <- overlapsAny(F, M)

cat("(1) PEAK-CENTRIC (each sex's own peaks, >=1bp overlap):\n")
cat(sprintf("    male   %5d = %5d unique + %5d overlapping-female\n", length(M), sum(!m_hit), sum(m_hit)))
cat(sprintf("    female %5d = %5d unique + %5d overlapping-male\n", length(F), sum(!f_hit), sum(f_hit)))
cat(sprintf("    NB shared is ASYMMETRIC (%d vs %d) - expected, overlap is many-to-many.\n\n",
            sum(m_hit), sum(f_hit)))

write_bed(M[!m_hit], "CC_male_unique_peakCentric.bed")
write_bed(F[!f_hit], "CC_female_unique_peakCentric.bed")
write_bed(M[m_hit],  "CC_male_shared_peakCentric_maleCoords.bed")
write_bed(F[f_hit],  "CC_female_shared_peakCentric_femaleCoords.bed")

## ---- (2) REGION-CENTRIC merge -> disjoint categories -----------------
## bedtools cat male.bed female.bed | bedtools sort | bedtools merge
u   <- reduce(c(granges(M), granges(F)))
inM <- overlapsAny(u, M)
inF <- overlapsAny(u, F)

male_only   <- u[ inM & !inF]
shared      <- u[ inM &  inF]
female_only <- u[!inM &  inF]

stopifnot(length(male_only) + length(shared) + length(female_only) == length(u))

cat("(2) REGION-CENTRIC (merged union; disjoint, sums exactly) <- used for the Venn:\n")
cat(sprintf("    male-only %d + shared %d + female-only %d = %d union regions\n",
            length(male_only), length(shared), length(female_only), length(u)))
cat(sprintf("    male circle  = %d   female circle = %d\n\n",
            length(male_only) + length(shared), length(female_only) + length(shared)))

write_bed(male_only,   "CC_male_only_regions.bed")
write_bed(female_only, "CC_female_only_regions.bed")
write_bed(shared,      "CC_shared_regions.bed")

## ---- Summary table ---------------------------------------------------
summ <- data.frame(
  Method   = c(rep("peak-centric", 4), rep("region-centric", 4)),
  Set      = c("male total", "male unique", "female total", "female unique",
               "male-only", "shared", "female-only", "union total"),
  N        = c(length(M), sum(!m_hit), length(F), sum(!f_hit),
               length(male_only), length(shared), length(female_only), length(u))
)
write.csv(summ, file.path(out_dir, "CC_peak_overlap_summary.csv"), row.names = FALSE)

## ---- Venn (region-centric) ------------------------------------------
## Font sizes: counts ~17pt, category labels ~18pt, title ~22pt on a 12pt device.
draw_venn <- function() {
grid.newpage()
vp <- draw.pairwise.venn(
  area1      = length(male_only) + length(shared),
  area2      = length(female_only) + length(shared),
  cross.area = length(shared),
  category   = c("Male", "Female"),
  fill       = c("#4C7FB8", "#C4622D"),
  alpha      = c(0.55, 0.55),
  lwd        = 2.5,
  col        = c("#2E4F73", "#7A3B18"),
  cex        = 1.4,     # the three counts
  cat.cex    = 1.5,     # "Male" / "Female"
  cat.col    = c("#2E4F73", "#7A3B18"),
  cat.pos    = c(-30, 30),
  cat.dist   = 0.05,
  fontfamily     = "sans",
  cat.fontfamily = "sans",
  cat.fontface   = "bold",
  margin     = 0.10,
  ind        = FALSE
)
## comma-separate the three counts (grobs carry raw labels e.g. "8204")
lab_map <- setNames(format(c(length(male_only), length(shared), length(female_only)), big.mark = ",", trim = TRUE),
                    as.character(c(length(male_only), length(shared), length(female_only))))
vp <- lapply(vp, function(g) {
  if (inherits(g, "text") && !is.null(g$label) && as.character(g$label) %in% names(lab_map))
    g$label <- lab_map[[as.character(g$label)]]
  g
})
class(vp) <- "gList"

grid.draw(vp)
grid.text("Egr1 Calling-Card peak overlap",
          y = 0.955, gp = gpar(fontsize = 18, fontface = "bold", fontfamily = "sans"))
grid.text("peaks called separately per sex",
          y = 0.905, gp = gpar(fontsize = 14, fontfamily = "sans", col = "grey30"))
grid.text(sprintf("merged union regions, ≥ 1 bp overlap  |  n = %s total", format(length(u), big.mark = ",")),
          y = 0.045, gp = gpar(fontsize = 14, fontfamily = "sans", col = "grey30"))
}

venn_pdf <- file.path(out_dir, "SupplFig4b_CC_peak_overlap_Venn.pdf")
venn_png <- file.path(out_dir, "SupplFig4b_CC_peak_overlap_Venn.png")
pdf(venn_pdf, width = 7.5, height = 7);                      draw_venn(); dev.off()
png(venn_png, width = 7.5, height = 7, units = "in", res = 200, pointsize = 12); draw_venn(); dev.off()

cat("Wrote:\n  ", venn_pdf, "\n  ", venn_png, "\n  ",
    file.path(out_dir, "CC_peak_overlap_summary.csv"), "\n  + 7 bed files in ", out_dir, "\n", sep = "")

File: 04_callingcard_analysis/CC_peak_overlap_venn.py
"""
Male vs Female Calling-Card peak overlap -> Venn + peak sets   [CORRECTED]

WHAT WAS WRONG WITH THE ORIGINAL NOTEBOOK CELL
----------------------------------------------
The original used:
    common       = peak_data_Male_gr.intersect(peak_data_Female_gr)
    unique_peaks1 = peak_data_Male_gr.subtract(peak_data_Female_gr)
    unique_peaks2 = peak_data_Female_gr.subtract(peak_data_Male_gr)
on inputs `peak_annotation_Male_20000` / `peak_annotation_Female_20000`.

Three independent problems:

1. WRONG INPUT. `*_20000` is the 20 kb-nearest-gene-FILTERED set (male 9,456 /
   female 9,603), not the raw calls (male 11,977 / female 11,772). Peaks >20 kb
   from any gene were silently excluded from the Venn. The 20 kb filter belongs to
   the CC-vs-DE regulon step; it has nothing to do with whether a peak is shared
   between sexes.

2. `.intersect()` and `.subtract()` are INTERVAL ARITHMETIC, not peak classification.
     - `.intersect()` returns one row per overlapping PAIR, CLIPPED to the overlap
       segment. So "shared" coordinates are intersection segments, not real peaks,
       and one male peak overlapping two female peaks yields TWO rows.
     - `.subtract()` returns the LEFTOVER FRAGMENTS after erasing the overlap. A
       peak that partly overlaps the other sex is not removed — it is trimmed, and
       can even be SPLIT into two fragments, each counted as a "unique peak".
   The counts are therefore fragment counts, not peak counts, and the categories
   do not partition the input.

3. The counts cannot be summed into a Venn. `venn2(subsets=(u1, u2, common))`
   assumes the three are disjoint counts of the SAME kind of object. Here they are
   fragments produced by three different operations.

THE FIX
-------
Use whole-peak classification (`.overlap()` / `.overlap(invert=True)`, which are
`bedtools intersect -u` / `-v`), and report BOTH honest framings:

  (1) PEAK-CENTRIC  - keeps each sex's own peaks intact. The two "shared" counts
      DISAGREE (3,712 male-side vs 3,719 female-side) because overlap is
      many-to-many. This is correct and expected; it just cannot be drawn as a
      2-circle Venn.
  (2) REGION-CENTRIC - merge the union, then ask which sexes hit each region.
      Categories are disjoint and sum exactly. THIS is what the Venn shows.

Requires: pandas, pyranges, matplotlib, matplotlib-venn
"""

import os
import pandas as pd
import pyranges as pr
import matplotlib.pyplot as plt
from matplotlib_venn import venn2

# ---- Paths -----------------------------------------------------------------
CC_DIR = r"C:\Users\loril\Documents\Egr1\Egr1CC_vs_Egr1KDBulkRNA_FINAL_July2025"
OUT_DIR = r"C:\Users\loril\Documents\GitHub\Egr1-Llaci\04_callingcard_analysis\output\peak_overlap"
os.makedirs(OUT_DIR, exist_ok=True)

MALE_RAW = os.path.join(CC_DIR, "Male_Egr1CC_peaks_072825.txt")
FEMALE_RAW = os.path.join(CC_DIR, "Female_Egr1CC_peaks_072825.txt")


def load_peaks(path):
    """Raw MACCs calls -> PyRanges. NOTE: the raw file, not the *_20000 filtered one."""
    df = pd.read_csv(path, sep="\t")
    df = df.rename(columns={"Chr": "Chromosome"})[["Chromosome", "Start", "End"]]
    return pr.PyRanges(df)


male = load_peaks(MALE_RAW)
female = load_peaks(FEMALE_RAW)
print(f"Raw peaks:  male {len(male)}   female {len(female)}")

# ---- (1) PEAK-CENTRIC : bedtools intersect -u / -v --------------------------
# .overlap() keeps WHOLE peaks that overlap (unlike .intersect(), which clips).
male_shared = male.overlap(female)
male_unique = male.overlap(female, invert=True)
female_shared = female.overlap(male)
female_unique = female.overlap(male, invert=True)

print("\n(1) PEAK-CENTRIC (each sex's own peaks, >=1 bp overlap):")
print(f"    male   {len(male):5d} = {len(male_unique):5d} unique + {len(male_shared):5d} overlapping-female")
print(f"    female {len(female):5d} = {len(female_unique):5d} unique + {len(female_shared):5d} overlapping-male")
print(f"    NB shared is ASYMMETRIC ({len(male_shared)} vs {len(female_shared)}) - expected, overlap is many-to-many.")

for gr, name in [(male_unique, "CC_male_unique_peakCentric"),
                 (female_unique, "CC_female_unique_peakCentric"),
                 (male_shared, "CC_male_shared_peakCentric_maleCoords"),
                 (female_shared, "CC_female_shared_peakCentric_femaleCoords")]:
    gr.df[["Chromosome", "Start", "End"]].to_csv(
        os.path.join(OUT_DIR, name + ".bed"), sep="\t", header=False, index=False)

# ---- (2) REGION-CENTRIC : merge union -> disjoint categories -----------------
union = pr.concat([male, female]).merge()
in_male = union.count_overlaps(male).df["NumberOverlaps"] > 0
in_female = union.count_overlaps(female).df["NumberOverlaps"] > 0

udf = union.df
male_only = udf[in_male & ~in_female]
shared = udf[in_male & in_female]
female_only = udf[~in_male & in_female]

assert len(male_only) + len(shared) + len(female_only) == len(udf), "categories must partition the union"

print("\n(2) REGION-CENTRIC (merged union; disjoint, sums exactly) <- used for the Venn:")
print(f"    male-only {len(male_only)} + shared {len(shared)} + female-only {len(female_only)} = {len(udf)} union regions")
print(f"    male circle = {len(male_only) + len(shared)}   female circle = {len(female_only) + len(shared)}")

for df_, name in [(male_only, "CC_male_only_regions"),
                  (shared, "CC_shared_regions"),
                  (female_only, "CC_female_only_regions")]:
    df_[["Chromosome", "Start", "End"]].to_csv(
        os.path.join(OUT_DIR, name + ".bed"), sep="\t", header=False, index=False)

pd.DataFrame({
    "Method": ["peak-centric"] * 4 + ["region-centric"] * 4,
    "Set": ["male total", "male unique", "female total", "female unique",
            "male-only", "shared", "female-only", "union total"],
    "N": [len(male), len(male_unique), len(female), len(female_unique),
          len(male_only), len(shared), len(female_only), len(udf)],
}).to_csv(os.path.join(OUT_DIR, "CC_peak_overlap_summary.csv"), index=False)

# ---- Venn (region-centric) --------------------------------------------------
fig, ax = plt.subplots(figsize=(7.5, 7))
v = venn2(subsets=(len(male_only), len(female_only), len(shared)),
          set_labels=("Male", "Female"), ax=ax)

for sid, colour in {"10": "#4C7FB8", "01": "#C4622D", "11": "#8A5344"}.items():
    p = v.get_patch_by_id(sid)
    if p:
        p.set_color(colour)
        p.set_edgecolor("black")
        p.set_linewidth(2)
        p.set_alpha(0.55)

# Font sizes per project figure rules: counts/labels >=14 pt, title >=18 pt.
for sid in ("10", "01", "11"):
    lbl = v.get_label_by_id(sid)
    if lbl:
        lbl.set_fontsize(17)
        lbl.set_text(f"{int(lbl.get_text()):,}")   # thousands separators
for lbl in v.set_labels:
    if lbl:
        lbl.set_fontsize(18)
        lbl.set_fontweight("bold")

ax.set_title("Egr1 Calling-Card peak overlap", fontsize=18, fontweight="bold", pad=28)
ax.text(0.5, 1.02, "peaks called separately per sex", transform=ax.transAxes,
        ha="center", fontsize=14, color="grey")
ax.text(0.5, -0.04, f"merged union regions, >=1 bp overlap  |  n = {len(udf):,} total",
        transform=ax.transAxes, ha="center", fontsize=14, color="grey")

plt.tight_layout()
for ext in ("pdf", "png"):
    plt.savefig(os.path.join(OUT_DIR, f"SupplFig4b_CC_peak_overlap_Venn_py.{ext}"),
                dpi=300, bbox_inches="tight")
print(f"\nWrote figure + 7 beds + summary to:\n  {OUT_DIR}")
